### ISSUE TITLE
Prometheus Alertmanager shows ntpd-rs down intermittently

### ISSUE THREAD

**Issue by Bradatletsencrypt** [pendulum-project/ntpd-rs#1497]
Hello,

We have been working on deploying `ntpd-rs` and it's been going well. We really appreciate that we are able to get metrics in a way that's very easy and intuitive. Upon deploying to staging and then production, we noticed that sometimes alertmanager had a pending alert for `ntpd-rs` being down. It never fired and would always clear on the next scrape. I decided to see if I could see this same behavior in my dev environment, and I am able to:

![image](https://github.com/pendulum-project/ntpd-rs/assets/132489699/5900f1b9-07d5-4d21-8eb4-5ad2c6907ba9)

When investigating the nodes that report down, systemd doesn't show any restarts of the `ntpd-rs` service or the related metrics service. I enabled debug logging in alertmanager and grepped through logs by the node names that appeared down only to not find any log entries for them. For example, I can see scrape targets that have been reported down with:
```
sudo journalctl -u prometheus-alertmanager --since "12 hours ago" --until "now" | grep ntpd-rs | grep Down | cut -d, -f2 | cut -d\" -f2 | sed -e 's/\\//g' | sort | uniq -c
```
The only output from this is the scrape targets I did _not_ bring up in my dev environment, so that's expected to be in the log. Grepping for the node names alertmanager thought were down for a scrape yields nothing in the logs. I then wanted to see if I could figure out what was actually happening during one of these "failures" -- for this I set up a couple of different loops:

```
# capture the time it takes to cURL each scrape target
for i in {1..500}; do for i in $(cat hosts.txt); do echo $i; time curl -s $i:9975/metrics | wc -l; done >> time_test.log 2>&1; done

# count the number of lines in output of each scrape target
for i in {1..250}; do for i in $(cat hosts.txt); do echo $i; curl -s $i:9975/metrics | wc -l; done; sleep 5; done >> wc_test.log
```
For the former, I can just grep for `real` to see if any large timeouts appear to be happening:
```
$ grep real time_test.log 
real	0m0.010s
real	0m0.013s
real	0m0.008s
real	0m0.008s
real	0m0.009s
real	0m0.007s
real	0m0.008s
real	0m0.008s
real	0m0.007s
real	0m0.008s
real	0m0.009s
real	0m0.012s
real	0m0.009s
real	0m0.012s
real	0m0.010s
real	0m0.008s
real	0m0.008s
real	0m0.012s
real	0m0.011s
real	0m0.011s
real	0m0.010s
real	0m0.012s
real	0m0.012s
real	0m0.010s
[...more like this...]
```
For the latter I'm just looking for any response that doesn't have the same output from `wc -l` as the other lines, they are all identical and are not indicative of a problem with the output.

A colleague who has more experience digging through rust was also unsure what to make of this, though he did point to this line as a potential place to start looking: https://github.com/pendulum-project/ntpd-rs/blob/79a6f38c12f25e39e571d77e61f2172fd611b2f6/ntpd/src/daemon/sockets.rs#L15

He theorized on Slack:
> The protocol seems to be that the exporter connects, the daemon writes a JSON object to the socket, and then the exporter does one receive syscall, and parses the result. My worry is that, if the message exceeds some buffer length, the partial response wouldn't parse, and the receiving end would stop there. Ideally the protocol should be length prefixed or delimited somehow, but failing that, reading more and retrying would work and remain backwards compatible.

This is a bit beyond my experience in reading rust, but I'm including it as a place that might be helpful to start looking.

I also didn't see much in the `ntpd-rs-metrics` logs. I see this line sometimes:
```
Jun 12 20:34:51 [HOSTNAME] ntp-metrics-exporter[54319]: 2024-06-12T20:34:51.103221Z  WARN ntpd::metrics::exporter: hit an error: No such file or directory (os error 2)
```
I checked each machine in my dev environment and this message does not appear in any logs today, while I do see a few hits on the graph showing ntpd-rs down today. So I do not think they are related and had more to do with the initial rollout into my dev environment yesterday.

I have observed this in our production, staging and dev environments. I do not think this issue stops us from continuing our rollout as we've never had one of these alerts actually fire, but we would very much like to understand what we're seeing here. I'm happy to grab any other details you need, just let me know.

Thanks!
Brad

**Comment by cikzh** [comment]
We should check if the same fix is needed in [Statime](https://github.com/pendulum-project/statime)

**Comment by cikzh** [comment]
Possibly related to #1496

**Comment by rnijveld** [comment]
We've just released [1.2.0](https://github.com/pendulum-project/ntpd-rs/releases/tag/v1.2.0), which includes a change in the way metrics connections are handled, hopefully improving the responsiveness, especially whenever the daemon is particularly busy. Let me know if that changes any of the behavior you are seeing.

**Comment by Bradatletsencrypt** [comment]
Thank you for the update! We have been watching the 1.2.0 release and are looking forward to getting to work testing in the coming days.

**Comment by mcpherrinm** [comment]
![image](https://github.com/user-attachments/assets/3acde646-706f-49ba-aaf5-8dbcc299da35)
It does seem like it's still happening - I upgraded to 1.2.0 at 17:00 in this image

**Comment by mcpherrinm** [comment]
I've turned on debug level logs in prometheus, and captured the failures:

```
Jul 20 00:29:02 metrics02 prometheus[143493]: 2024/07/20 00:29:02 Unsolicited response received on idle HTTP channel starting with "HTTP/1.1 200 OK\r\ncontent-type: text/plain\r\ncontent-length: 5863\r\n\r\n# HELP ntp_uptime_seconds Time that the ntp daemon is running.\n# TYPE ntp_uptime_seconds gauge\n# UNIT ntp_uptime_seconds seconds\nntp_uptime_seconds{version=\"1.2.0\",build_commit=\"ed52889ab4517c70fb8c1481e983666e2033b15b\",build_commit_date=\"2024-07-11\"} 208147.229675719\n# HELP ntp_system_poll_interval_seconds [DEPRECATED] Time between polls of the system.\n# TYPE ntp_system_poll_interval_seconds gauge\n# UNIT ntp_system_poll_interval_seconds seconds\nntp_system_poll_interval_seconds 32.00000000745058\n# HELP ntp_system_accumulated_steps_seconds Accumulated amount of seconds that the system needed to jump the time.\n# TYPE ntp_system_accumulated_steps_seconds gauge\n# UNIT ntp_system_accumulated_steps_seconds seconds\nntp_system_accumulated_steps_seconds 27.78465242818572\n# HELP ntp_system_accumulated_steps_threshold_seconds Threshold for the accumulated step amount at which the NTP daemon will exit (or -1 if no threshold was set).\n# TYPE ntp_system_accumulated_steps_threshold_seconds gauge\n# UNIT ntp_system_accumulated_steps_threshold_seconds seconds\nntp_system_accumulated_steps_threshold_seconds -1\n# HELP ntp_system_leap_indicator Indicates that a leap second will take place.\n# TYPE ntp_system_leap_indicator gauge\nntp_system_leap_indicator 0\n# HELP ntp_system_root_delay_seconds Distance to the closest root time source.\n# TYPE ntp_system_root_delay_seconds gauge\n# UNIT ntp_system_root_delay_seconds seconds\nntp_system_root_delay_seconds 0.011370129885936652\n# HELP ntp_system_root_dispersion_seconds Estimate of how precise our time is.\n# TYPE ntp_system_root_dispersion_seconds gauge\n# UNIT ntp_system_root_dispersion_seconds seconds\nntp_system_root_dispersion_seconds 0.00011984515006650359\n# HELP ntp_system_stratum Stratum of our clock.\n# TYPE ntp_system_stratum gauge\nntp_system_stratum 4\n# HELP ntp_source_poll_interval_seconds Time between polls of the source.\n# TYPE ntp_source_poll_interval_seconds gauge\n# UNIT ntp_source_poll_interval_seconds seconds\nntp_source_poll_interval_seconds{name=\"10.4.11.158:123\",address=\"10.4.11.158:123\",id=\"4137\"} 32.00000000745058\nntp_source_poll_interval_seconds{name=\"10.4.11.157:123\",address=\"10.4.11.157:123\",id=\"1\"} 128.00000002980232\n# HELP ntp_source_unanswered_polls Number of polls since the last succesful poll with a maximum of eight.\n# TYPE ntp_source_unanswered_polls gauge\nntp_source_unanswered_polls{name=\"10.4.11.158:123\",address=\"10.4.11.158:123\",id=\"4137\"} 0\nntp_source_unanswered_polls{name=\"10.4.11.157:123\",address=\"10.4.11.157:123\",id=\"1\"} 0\n# HELP ntp_source_offset_seconds Offset between the upstream source and system time.\n# TYPE ntp_source_offset_seconds gauge\n# UNIT ntp_source_offset_seconds seconds\nntp_source_offset_seconds{name=\"10.4.11.158:123\",address=\"10.4.11.158:123\",id=\"4137\"} 0.00007596658544521001\nntp_source_offset_seconds{name=\"10.4.11.157:123\",address=\"10.4.11.157:123\",id=\"1\"} 0.0002724614460655631\n# HELP ntp_source_delay_seconds Current round-trip delay to the upstream source.\n# TYPE ntp_source_delay_seconds gauge\n# UNIT ntp_source_delay_seconds seconds\nntp_source_delay_seconds{name=\"10.4.11.158:123\",address=\"10.4.11.158:123\",id=\"4137\"} 0.00025961850775862544\nntp_source_delay_seconds{name=\"10.4.11.157:123\",address=\"10.4.11.157:123\",id=\"1\"} 0.00021939608273547983\n# HELP ntp_source_uncertainty_seconds Estimated error of the source clock.\n# TYPE ntp_source_uncertainty_seconds gauge\n# UNIT ntp_source_uncertainty_seconds seconds\nntp_source_uncertainty_seconds{name=\"10.4.11.158:123\",address=\"10.4.11.158:123\",id=\"4137\"} 0.000008061993869035969\nntp_source_uncertainty_seconds{name=\"10.4.11.157:123\",address=\"10.4.11.157:123\",id=\"1\"} 0.00000890577212183405\n# HELP ntp_source_root_delay_seconds Root delay reported by the time source.\n# TYPE ntp_source_root_delay_seconds gauge\n# UNIT ntp_source_root_delay_seconds seconds\nntp_source_root_delay_seconds{name=\"10.4.11.158:123\",address=\"10.4.11.158:123\",id=\"4137\"} 0.011276245119812955\nntp_source_root_delay_seconds{name=\"10.4.11.157:123\",address=\"10.4.11.157:123\",id=\"1\"} 0.01"; err=<nil>
Jul 20 00:29:02 metrics02 prometheus[143493]: ts=2024-07-20T00:29:02.919Z caller=scrape.go:1372 level=debug component="scrape manager" scrape_pool=ntpd-rs target=http://metrics02:9975/metrics msg="Scrape failed" err="Get \"http://metrics02:9975/metrics\": readLoopPeekFailLocked: %!w(<nil>)"
Jul 20 00:48:52 metrics02 prometheus[143493]: 2024/07/20 00:48:52 Unsolicited response received on idle HTTP channel starting with "HTTP/1.1 200 OK\r\ncontent-type: text/plain\r\ncontent-length: 5867\r\n\r\n# HELP ntp_uptime_seconds Time that the ntp daemon is running.\n# TYPE ntp_uptime_seconds gauge\n# UNIT ntp_uptime_seconds seconds\nntp_uptime_seconds{version=\"1.2.0\",build_commit=\"ed52889ab4517c70fb8c1481e983666e2033b15b\",build_commit_date=\"2024-07-11\"} 209337.230171202\n# HELP ntp_system_poll_interval_seconds [DEPRECATED] Time between polls of the system.\n# TYPE ntp_system_poll_interval_seconds gauge\n# UNIT ntp_system_poll_interval_seconds seconds\nntp_system_poll_interval_seconds 32.00000000745058\n# HELP ntp_system_accumulated_steps_seconds Accumulated amount of seconds that the system needed to jump the time.\n# TYPE ntp_system_accumulated_steps_seconds gauge\n# UNIT ntp_system_accumulated_steps_seconds seconds\nntp_system_accumulated_steps_seconds 27.78465242818572\n# HELP ntp_system_accumulated_steps_threshold_seconds Threshold for the accumulated step amount at which the NTP daemon will exit (or -1 if no threshold was set).\n# TYPE ntp_system_accumulated_steps_threshold_seconds gauge\n# UNIT ntp_system_accumulated_steps_threshold_seconds seconds\nntp_system_accumulated_steps_threshold_seconds -1\n# HELP ntp_system_leap_indicator Indicates that a leap second will take place.\n# TYPE ntp_system_leap_indicator gauge\nntp_system_leap_indicator 0\n# HELP ntp_system_root_delay_seconds Distance to the closest root time source.\n# TYPE ntp_system_root_delay_seconds gauge\n# UNIT ntp_system_root_delay_seconds seconds\nntp_system_root_delay_seconds 0.011370129885936652\n# HELP ntp_system_root_dispersion_seconds Estimate of how precise our time is.\n# TYPE ntp_system_root_dispersion_seconds gauge\n# UNIT ntp_system_root_dispersion_seconds seconds\nntp_system_root_dispersion_seconds 0.00011984515006650359\n# HELP ntp_system_stratum Stratum of our clock.\n# TYPE ntp_system_stratum gauge\nntp_system_stratum 4\n# HELP ntp_source_poll_interval_seconds Time between polls of the source.\n# TYPE ntp_source_poll_interval_seconds gauge\n# UNIT ntp_source_poll_interval_seconds seconds\nntp_source_poll_interval_seconds{name=\"10.4.11.158:123\",address=\"10.4.11.158:123\",id=\"4137\"} 32.00000000745058\nntp_source_poll_interval_seconds{name=\"10.4.11.157:123\",address=\"10.4.11.157:123\",id=\"1\"} 128.00000002980232\n# HELP ntp_source_unanswered_polls Number of polls since the last succesful poll with a maximum of eight.\n# TYPE ntp_source_unanswered_polls gauge\nntp_source_unanswered_polls{name=\"10.4.11.158:123\",address=\"10.4.11.158:123\",id=\"4137\"} 0\nntp_source_unanswered_polls{name=\"10.4.11.157:123\",address=\"10.4.11.157:123\",id=\"1\"} 0\n# HELP ntp_source_offset_seconds Offset between the upstream source and system time.\n# TYPE ntp_source_offset_seconds gauge\n# UNIT ntp_source_offset_seconds seconds\nntp_source_offset_seconds{name=\"10.4.11.158:123\",address=\"10.4.11.158:123\",id=\"4137\"} -0.00006872205996623311\nntp_source_offset_seconds{name=\"10.4.11.157:123\",address=\"10.4.11.157:123\",id=\"1\"} 0.0001932848245355498\n# HELP ntp_source_delay_seconds Current round-trip delay to the upstream source.\n# TYPE ntp_source_delay_seconds gauge\n# UNIT ntp_source_delay_seconds seconds\nntp_source_delay_seconds{name=\"10.4.11.158:123\",address=\"10.4.11.158:123\",id=\"4137\"} 0.00026358384645161774\nntp_source_delay_seconds{name=\"10.4.11.157:123\",address=\"10.4.11.157:123\",id=\"1\"} 0.0002293309663025036\n# HELP ntp_source_uncertainty_seconds Estimated error of the source clock.\n# TYPE ntp_source_uncertainty_seconds gauge\n# UNIT ntp_source_uncertainty_seconds seconds\nntp_source_uncertainty_seconds{name=\"10.4.11.158:123\",address=\"10.4.11.158:123\",id=\"4137\"} 0.000013578450310411502\nntp_source_uncertainty_seconds{name=\"10.4.11.157:123\",address=\"10.4.11.157:123\",id=\"1\"} 0.000008862698452748055\n# HELP ntp_source_root_delay_seconds Root delay reported by the time source.\n# TYPE ntp_source_root_delay_seconds gauge\n# UNIT ntp_source_root_delay_seconds seconds\nntp_source_root_delay_seconds{name=\"10.4.11.158:123\",address=\"10.4.11.158:123\",id=\"4137\"} 0.011535644533935852\nntp_source_root_delay_seconds{name=\"10.4.11.157:123\",address=\"10.4.11.157:123\",id=\"1\"} 0.0"; err=<nil>
Jul 20 00:48:52 metrics02 prometheus[143493]: ts=2024-07-20T00:48:52.920Z caller=scrape.go:1372 level=debug component="scrape manager" scrape_pool=ntpd-rs target=http://metrics02:9975/metrics msg="Scrape failed" err="Get \"http://metrics02:9975/metrics\": readLoopPeekFailLocked: %!w(<nil>)"
Jul 20 00:56:42 metrics02 prometheus[143493]: 2024/07/20 00:56:42 Unsolicited response received on idle HTTP channel starting with "HTTP/1.1 200 OK\r\ncontent-type: text/plain\r\ncontent-length: 5869\r\n\r\n# HELP ntp_uptime_seconds Time that the ntp daemon is running.\n# TYPE ntp_uptime_seconds gauge\n# UNIT ntp_uptime_seconds seconds\nntp_uptime_seconds{version=\"1.2.0\",build_commit=\"ed52889ab4517c70fb8c1481e983666e2033b15b\",build_commit_date=\"2024-07-11\"} 209807.230353404\n# HELP ntp_system_poll_interval_seconds [DEPRECATED] Time between polls of the system.\n# TYPE ntp_system_poll_interval_seconds gauge\n# UNIT ntp_system_poll_interval_seconds seconds\nntp_system_poll_interval_seconds 32.00000000745058\n# HELP ntp_system_accumulated_steps_seconds Accumulated amount of seconds that the system needed to jump the time.\n# TYPE ntp_system_accumulated_steps_seconds gauge\n# UNIT ntp_system_accumulated_steps_seconds seconds\nntp_system_accumulated_steps_seconds 27.78465242818572\n# HELP ntp_system_accumulated_steps_threshold_seconds Threshold for the accumulated step amount at which the NTP daemon will exit (or -1 if no threshold was set).\n# TYPE ntp_system_accumulated_steps_threshold_seconds gauge\n# UNIT ntp_system_accumulated_steps_threshold_seconds seconds\nntp_system_accumulated_steps_threshold_seconds -1\n# HELP ntp_system_leap_indicator Indicates that a leap second will take place.\n# TYPE ntp_system_leap_indicator gauge\nntp_system_leap_indicator 0\n# HELP ntp_system_root_delay_seconds Distance to the closest root time source.\n# TYPE ntp_system_root_delay_seconds gauge\n# UNIT ntp_system_root_delay_seconds seconds\nntp_system_root_delay_seconds 0.011370129885936652\n# HELP ntp_system_root_dispersion_seconds Estimate of how precise our time is.\n# TYPE ntp_system_root_dispersion_seconds gauge\n# UNIT ntp_system_root_dispersion_seconds seconds\nntp_system_root_dispersion_seconds 0.00011984515006650359\n# HELP ntp_system_stratum Stratum of our clock.\n# TYPE ntp_system_stratum gauge\nntp_system_stratum 4\n# HELP ntp_source_poll_interval_seconds Time between polls of the source.\n# TYPE ntp_source_poll_interval_seconds gauge\n# UNIT ntp_source_poll_interval_seconds seconds\nntp_source_poll_interval_seconds{name=\"10.4.11.158:123\",address=\"10.4.11.158:123\",id=\"4137\"} 32.00000000745058\nntp_source_poll_interval_seconds{name=\"10.4.11.157:123\",address=\"10.4.11.157:123\",id=\"1\"} 128.00000002980232\n# HELP ntp_source_unanswered_polls Number of polls since the last succesful poll with a maximum of eight.\n# TYPE ntp_source_unanswered_polls gauge\nntp_source_unanswered_polls{name=\"10.4.11.158:123\",address=\"10.4.11.158:123\",id=\"4137\"} 0\nntp_source_unanswered_polls{name=\"10.4.11.157:123\",address=\"10.4.11.157:123\",id=\"1\"} 0\n# HELP ntp_source_offset_seconds Offset between the upstream source and system time.\n# TYPE ntp_source_offset_seconds gauge\n# UNIT ntp_source_offset_seconds seconds\nntp_source_offset_seconds{name=\"10.4.11.158:123\",address=\"10.4.11.158:123\",id=\"4137\"} -0.00012168870310338417\nntp_source_offset_seconds{name=\"10.4.11.157:123\",address=\"10.4.11.157:123\",id=\"1\"} 0.00015929574150575691\n# HELP ntp_source_delay_seconds Current round-trip delay to the upstream source.\n# TYPE ntp_source_delay_seconds gauge\n# UNIT ntp_source_delay_seconds seconds\nntp_source_delay_seconds{name=\"10.4.11.158:123\",address=\"10.4.11.158:123\",id=\"4137\"} 0.00025238562381183394\nntp_source_delay_seconds{name=\"10.4.11.157:123\",address=\"10.4.11.157:123\",id=\"1\"} 0.00023389770654819387\n# HELP ntp_source_uncertainty_seconds Estimated error of the source clock.\n# TYPE ntp_source_uncertainty_seconds gauge\n# UNIT ntp_source_uncertainty_seconds seconds\nntp_source_uncertainty_seconds{name=\"10.4.11.158:123\",address=\"10.4.11.158:123\",id=\"4137\"} 0.000007652910428040873\nntp_source_uncertainty_seconds{name=\"10.4.11.157:123\",address=\"10.4.11.157:123\",id=\"1\"} 0.000009204261007067808\n# HELP ntp_source_root_delay_seconds Root delay reported by the time source.\n# TYPE ntp_source_root_delay_seconds gauge\n# UNIT ntp_source_root_delay_seconds seconds\nntp_source_root_delay_seconds{name=\"10.4.11.158:123\",address=\"10.4.11.158:123\",id=\"4137\"} 0.011795043948058748\nntp_source_root_delay_seconds{name=\"10.4.11.157:123\",address=\"10.4.11.157:123\",id=\"1\"} 0"; err=<nil>
Jul 20 00:56:42 metrics02 prometheus[143493]: ts=2024-07-20T00:56:42.920Z caller=scrape.go:1372 level=debug component="scrape manager" scrape_pool=ntpd-rs target=http://metrics02:9975/metrics msg="Scrape failed" err="Get \"http://metrics02:9975/metrics\": readLoopPeekFailLocked: %!w(<nil>)"
```

**Comment by divergentdave** [comment]
Looking at the exporter source code, I notice that it starts writing a response soon after accepting a connection, and never reads from the socket. This sets up a race condition where Prometheus may get a response before it has sent its request. I think many HTTP client libraries will shut down a connection if there is any data to be read immediately before they send a request.

**Comment by rnijveld** [comment]
That was definitely what was going on! A previous version of this metrics exporter did check on the request, but somehow that was removed, and it never caused any issues with our own metrics tracking over a relatively slow network. On a fast network, the response could easily have been received before the request was sent (given how fast we can export metrics). I've added some code that does wait until at least all of the headers have been received (up to 2048 bytes, otherwise it is an error, but in my quick testing, such a set of HTTP headers is unreasonably large and unlikely to happen). While we still don't really need or use the headers, that at least should prevent any race conditions. I think that will finally solve the issue you are seeing!
